SpatialGDK/Source/SpatialGDK/Private/Utils/EntityFactory.cpp

```diff
@@ -83,108 +124,115 @@ TArray<FWorkerComponentData> EntityFactory::CreateEntityComponents(USpatialActor
 		{
 			return;
 		}
 
-		ComponentWriteAcl.Add(ComponentId, AuthoritativeWorkerRequirementSet);
+		Acl->ComponentWriteAcl.Add(ComponentId, AuthoritativeWorkerRequirementSet);
 	});
 
-	for (auto& SubobjectInfoPair : Info.SubobjectInfo)
+	// Add debugging utilities, if we are not compiling a shipping build
+#if !UE_BUILD_SHIPPING
+	Acl->ComponentWriteAcl.Add(SpatialConstants::GDK_DEBUG_COMPONENT_ID, AuthoritativeWorkerRequirementSet);
+	if (NetDriver->SpatialDebugger != nullptr)
 	{
-		const FClassInfo& SubobjectInfo = SubobjectInfoPair.Value.Get();
+		check(NetDriver->VirtualWorkerTranslator != nullptr);
 
-		// Static subobjects aren't guaranteed to exist on actor instances, check they are present before adding write acls
-		TWeakObjectPtr<UObject> Subobject = PackageMap->GetObjectFromUnrealObjectRef(FUnrealObjectRef(EntityId, SubobjectInfoPair.Key));
-		if (!Subobject.IsValid())
-		{
-			continue;
-		}
+		const PhysicalWorkerName* PhysicalWorkerName =
+			NetDriver->VirtualWorkerTranslator->GetPhysicalWorkerForVirtualWorker(IntendedVirtualWorkerId);
+		FColor InvalidServerTintColor = NetDriver->SpatialDebugger->InvalidServerTintColor;
+		FColor IntentColor =
+			PhysicalWorkerName != nullptr ? SpatialGDK::GetColorForWorkerName(*PhysicalWorkerName) : InvalidServerTintColor;
 
-		ForAllSchemaComponentTypes([&](ESchemaComponentType Type) {
-			Worker_ComponentId ComponentId = SubobjectInfo.SchemaComponents[Type];
-			if (ComponentId == SpatialConstants::INVALID_COMPONENT_ID)
-			{
-				return;
-			}
+		const bool bIsLocked = NetDriver->LockingPolicy->IsLocked(Actor);
 
-			ComponentWriteAcl.Add(ComponentId, AuthoritativeWorkerRequirementSet);
-		});
+		SpatialDebugging DebuggingInfo(SpatialConstants::INVALID_VIRTUAL_WORKER_ID, InvalidServerTintColor, IntendedVirtualWorkerId,
+									   IntentColor, bIsLocked);
+		EntityComps.ComponentDatas.Add(DebuggingInfo.CreateComponentData());
+		Acl->ComponentWriteAcl.Add(SpatialConstants::SPATIAL_DEBUGGING_COMPONENT_ID, AuthoritativeWorkerRequirementSet);
 	}
+#endif
+
+	for (const auto& Pair : EntityComps.ComponentsToDelegateToAuthoritativeWorker)
+	{
+		Acl->ComponentWriteAcl.Add(Pair.Key, AuthoritativeWorkerRequirementSet);
+	}
+
+	if (Actor->IsA<APlayerController>())
+	{
+		Acl->ReadAcl = AnyServerOrOwningClientRequirementSet;
+		// PRCOMMENT: This will overwrite the previous value, whatever it was.
+		// Not sure what's the best way here. If for some reason we create skeleton entities for player controllers, not sure who should see
+		// them. Maybe we should restrict to just servers if we do not have players yet?
+	}
+
+	// Add actual load balancing components
+	EntityComps.ComponentDatas.Add(NetOwningClientWorker(GetConnectionOwningWorkerId(Actor)).CreateComponentData());
+	EntityComps.ComponentDatas.Add(AuthorityIntent::CreateAuthorityIntentData(IntendedVirtualWorkerId));
+}
+
+void EntityFactory::WriteUnrealComponents(EntityComponents& EntityComps, USpatialActorChannel* Channel,
```

Comment:
I would really like to remove more dependencies here, but I was kind of defeated.
We check the PackageMap, we need the FRepLayouts, that kind of needs the NetDriver, etc.
I could split it up a bit more and make the individual functions have fewer dependencies, but not sure how much further we can go.

Another thing that could theoretically be useful is constructing a template of an actor, without peeking at runtime data, using an approach similar to SchemaGen (so it would have bogus data, but the correct structure), but can't justify the time spent on doing that. Can be convinced otherwise.